core/rs/core/src/tableinfo.rs

```diff
@@ -0,0 +1,261 @@
+use crate::c::CPointer;
+use crate::c::{crsql_ColumnInfo, crsql_TableInfo};
+use alloc::boxed::Box;
+use alloc::ffi::CString;
+use alloc::format;
+use alloc::vec;
+use alloc::vec::Vec;
+use core::ffi::c_char;
+use num_traits::ToPrimitive;
+use sqlite_nostd as sqlite;
+use sqlite_nostd::Connection;
+use sqlite_nostd::ResultCode;
+use sqlite_nostd::StrRef;
+
+/**
+ * Given a table name, return the table info that describes that table.
+ * TableInfo is a struct that represents the results
+ * of pragma_table_info, pragma_index_list, pragma_index_info on a given table
+ * and its indices as well as some extra fields to facilitate crr creation.
+ */
+pub fn pull_table_info(
+    db: *mut sqlite::sqlite3,
+    table: &str,
+    table_info: *mut *mut crsql_TableInfo,
+    err: *mut *mut c_char,
+) -> Result<ResultCode, ResultCode> {
+    let sql = format!("SELECT count(*) FROM pragma_table_info('{table}')");
+    let columns_len = match db.prepare_v2(&sql).and_then(|stmt| {
+        stmt.step()?;
+        stmt.column_int(0)?.to_usize().ok_or(ResultCode::ERROR)
+    }) {
+        Ok(count) => count,
+        Err(code) => {
+            err.set(&format!("Failed to find columns for crr -- {table}"));
+            return Err(code);
+        }
+    };
+
+    let sql = format!(
+        "SELECT \"cid\", \"name\", \"type\", \"notnull\", \"pk\"
+         FROM pragma_table_info('{table}') ORDER BY cid ASC"
+    );
+    let column_infos = match db.prepare_v2(&sql) {
+        Ok(stmt) => {
+            let mut cols: Vec<crsql_ColumnInfo> = vec![];
+
+            while stmt.step()? == ResultCode::ROW {
+                cols.push(crsql_ColumnInfo {
+                    type_: stmt.column_text(2).map_err(free_cols(&cols))?.into_c_ptr(),
+                    name: stmt.column_text(1).map_err(free_cols(&cols))?.into_c_ptr(),
+                    notnull: stmt.column_int(3)?,
+                    cid: stmt.column_int(0)?,
+                    pk: stmt.column_int(4)?,
+                });
+            }
+
+            if cols.len() != columns_len {
+                err.set(&"Number of fetched columns did not match expected number of columns");
+                return Err(free_cols(&cols)(ResultCode::ERROR));
```

Comment:
Because I intended `free_cols` to be a higher order function used with `.map_err`.
```rs
// to make one-liner like this
.map_err(free_cols(&cols))
// not like this
.map_err(|err| {
  free_cols(&cols));
  return err;
})
```

Reply:
gotcha